File: app/src/main/java/com/tomclaw/appsend/core/StoreServiceHolder.java
package com.tomclaw.appsend.core;

import static com.tomclaw.appsend.core.ConfigKt.HOST_URL;

import com.tomclaw.appsend.di.legacy.LegacyInjector;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by solkin on 23.09.17.
 */
public class StoreServiceHolder {

    LegacyInjector injector;

    private StoreService service = null;

    private StoreService lazyInit() {
        if (service != null) return service;

        Retrofit retrofit = new Retrofit.Builder()
                .client(injector.httpClient)
                .baseUrl(HOST_URL + "/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        service = retrofit.create(StoreService.class);
        return service;
    }

    public StoreService getService() {
        return lazyInit();
    }
}

File: app/src/main/java/com/tomclaw/appsend/core/StoreServiceHolder.kt
package com.tomclaw.appsend.core

import com.tomclaw.appsend.di.legacy.LegacyInjector
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

/**
 * Created by solkin on 23.09.17.
 */
class StoreServiceHolder private constructor() {

    private var injector: LegacyInjector? = null
    private var service: StoreService? = null

    private fun init() {
        this.injector = LegacyInjector.getInstance()
    }

    private fun lazyInit(): StoreService {
        val injector = injector ?: throw Exception("LegacyInjector is not initialized yet")
        return service ?: run {
            Retrofit.Builder()
                .client(injector.httpClient)
                .baseUrl("$HOST_URL/")
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(StoreService::class.java).apply {
                    service = this
                }
        }
    }

    fun getService(): StoreService {
        return lazyInit()
    }

    companion object {
        private var instance: StoreServiceHolder? = null

        @JvmStatic
        fun getInstance(): StoreServiceHolder {
            return instance ?: run {
                StoreServiceHolder().apply { init() }
            }
        }
    }

}
